### DISCUSSION TITLE
vcpkg should have a means of providing reasonable time limits to build ports

### DISCUSSION THREAD

**Discussion by BillyONeal** [microsoft/vcpkg#17810] [Ideas]
I am told by @dan-shaw that sometimes tensorflow hangs during its build for some reason, and have seen qt similarly hang. We should be able to set a reasonable cap to prevent these ports from running out of control (either in the 'ci' command or generally). Say 'no port should take longer than 1 hour to build' or similar, and skip any constantly offending ports.

**Proposed solution**
An extra command that lists a timeout for all ports on the command line, or some way to declare timeouts in the CONTROL/manifest file.

**Comment by invy** [comment]
I think it should be done on CI and not in the package manager. What if there will be port which will require one hour. Anyway, why one hour? If somebody has overloaded CI, he'll get MUCH higher probability of false positives.
You just have to investigate why 'build' actually hangs. On windows core 2019 LSTC with hyper-v isolation it just happend for me, that the whole docker container just hang, but it's nothing can be addressed in vcpkg and package manager.

As a compromise for 'advanced users' command line option --build-timeout could be introduced, but still, I think it's anti-feature.

**Comment by Neumann-A** [comment]
>and have seen qt similarly hang

Probably a problem with jom......switching to nmake helps here.  https://github.com/microsoft/vcpkg/pull/8524

**Comment by BillyONeal** [comment]
@invy: We can't handle it in CI because in the worst case CI needs to build every port which takes some 14 hours. We don't want to burn 14 hours of VM time because a port that normally takes 10 seconds to build got stuck. We would only need to do this for ports that are problematic / have a history of getting hung and could easily set them much higher than the time they normally take.

**Comment by invy** [comment]
@BillyONeal I think 'reasonable' compromise would be to provide 'overlay' manifest file which would contain ports with corresponding timeouts, which you then pass as an argument. Making it the default behavior is just workaround which won't work for everybody and may cause headache for some developers.
I.e.
```
vcpkg install MyProblematicPort1 MyProblematicPort2 MyProblematicPort3 --timeout=1h # time-out for all ports
vcpkg install MyProblematicPort1 MyProblematicPort2 MyProblematicPort3 --timeout-list=build_timeouts.txt # detailed timeout information
```
build_timeouts.txt
```
MyProblematicPort1=1h
MyProblematicPort2=30m
MyProblematicPort3=2h
```

**Comment by BillyONeal** [comment]
@invy  A port build hanging forever is just as amounts
annoying for customers as it is for us, there's no reason they don't also deserve that kind of protection.
Again, we would only need to apply this to the 2-3 ports that have a history of occasionally getting stuck.
